src/orderly/cli/__init__.py

```diff
@@ -0,0 +1,204 @@
+@click.option(
+    "-p",
+    "--parameter",
+    nargs=2,
+    type=(str, str),
+    multiple=True,
+    help="Pass a string parameter to the report",
+)
+@click.option(
+    "-n",
+    "--numeric-parameter",
+    nargs=2,
+    type=(str, NumericalParamType()),
+    multiple=True,
+    help="Pass a numerical parameter to the report",
+)
+@click.option(
+    "-b",
+    "--bool-parameter",
+    nargs=2,
+    type=(str, bool),
+    multiple=True,
+    help="Pass a boolean parameter to the report",
+)
+@with_search_options
+@with_root
+def run(
+    name,
+    root,
+    parameter,
+    numeric_parameter,
+    bool_parameter,
+    search_options,
+):
+    """
+    Run a report.
+
+    Parameters to the report may be specified using the -p/-n/-b options,
+    depending on the parameter's type. These options may be specified multiple
+    times.
+    """
+    all_parameters = parameter + numeric_parameter + bool_parameter
+    counts = Counter(key for (key, _) in all_parameters)
+    duplicates = [key for key, n in counts.items() if n > 1]
+    if duplicates:
+        click.echo(
+            f"Parameters were specified multiple times: {format_list(duplicates)}",
+            err=True,
+        )
+        sys.exit(1)
+
+    id = orderly_run(
+        name,
+        root=root,
+        parameters=dict(all_parameters),
+        search_options=search_options,
+    )
+    click.echo(id)
+
+
+@cli.command()
+@click.argument("query")
+@with_search_options
+@with_root
+def search(query, root, search_options):
+    """
+    Search existing packets by query.
+
+    By default, the search will only return packets present in the local
+    repository. The --allow-remote option allows packets that are known locally
+    but only present remotely to also be returned.
+    """
+    packets = outpack.search.search(query, root=root, options=search_options)
+    if packets:
+        for id in packets:
+            click.echo(id)
+    else:
+        click.echo("No packets matching the query were found", err=True)
+        sys.exit(1)
+
+
+@cli.group()
+@with_root(propagate=True)
+def location():
+    """Manage remote locations."""
+    pass
+
+
+@location.command("list")
+@with_root
+def location_list(root):
+    """List configured remote locations."""
+    locations = outpack_location_list(root=root)
+    for name in locations:
+        click.echo(name)
+
+
+@location.command("rename")
+@click.argument("old")
+@click.argument("new")
+@with_root
+def location_rename(root, old, new):
+    """Rename a remote location."""
+    outpack_location_rename(old, new, root=root)
+
+
+@location.command("remove")
+@click.argument("name")
+@with_root
+def location_remove(root, name):
+    """Remove a remote location."""
+    outpack_location_remove(name, root=root)
+
+
+@location.command("add")
+@click.argument("name")
+@click.argument("type", type=click.Choice(["path", "ssh"]), metavar="TYPE")
+@click.argument("location")
+@with_root
+def location_add(root, name, type, location):
```

Comment:
At the moment this looks like `orderly location add foo path /path/to/repo`. We could possibly infer the type from the third argument, similar to what Git does, and it easily extends to an http implementation too.

This is the Git implementation: https://github.com/git/git/blob/83f1add914c6b4682de1e944ec0d1ac043d53d78/connect.c#L1070

It's an open question whether we'll ever have more types where this doesn't work as well, or even if we'll have types that need more than just a string value.

Reply:
so far we have:

* file (reasonably unique, could be disambiguated by `file://` of course)
* http (there are actually two of these - one for packit and one for outpack server, but they differ only in auth and will start with `https://` and `http://` respectively)
* ssh (could use `user@host:path` or make that `ssh://user@host:path` to be totally clear perhaps)

I doubt we'll have many more, and we're largely in control (via the schema) so we could definitely explore this